Hard(CoreData)

EugChesn opened the pull request.

Commit:
UI and Network

Commit:
Fix UI and some refactor style code

Commit:
Fix review

Commit:
Additional complete

Commit:
Add CoreData

Commit:
Add swiftlint

Commit:
Separate database and network; Add spinner

Commit:
Fix behavior ui

Commit:
Refactor ui and using cashe data(determinate offline mode)

Commit:
Refactor model(add callback)d

Commit:
fix some bugs

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+        tableView.dataSource = self
+    }
+    private func registerTableViewCells() {
+        let cell = UINib(nibName: "CustomDetailTableViewCell", bundle: nil)
```
Перенеси `CustomDetailTableViewCell` в блок с константами

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+import UIKit
+
+class DetailsController: UIViewController {
+    @IBOutlet weak var tableView: UITableView!
```
Визуально разделяй типы переменных, функции. Пусть будут лишние пустые строки (не больше 1), но читать будет удобнее и будет красивее.

`IBOutlet` ...

`var/let` ...

`func` ...

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+        let cell = UINib(nibName: "CustomDetailTableViewCell", bundle: nil)
+        tableView.register(cell, forCellReuseIdentifier: "DetailCell")
+    }
+    func sendData(_ anyData: Any?) {
```
Между закрывающейся скобкой и новым методом также пустую бы строку добавить.

EugChesn replied:
Все что связано с оформлением кода это дефолтные правила в swiftlint

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+
+extension DetailsController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return 50
```
Константа

EugChesn replied:
fix

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+        label.frame = CGRect.init(x: 15, y: 5, width: headerView.frame.width-10, height: headerView.frame.height-10)
+        label.text = "Info"
+        label.font = UIFont(name: "", size: 6) // my custom font
+        label.textColor = UIColor .systemBlue // my custom colour
```
Ты уверен, что это компилируется? :)

EugChesn replied:
Да))

a-mkrv commented on TestView/Controllers/DetailsController.swift:
```diff
@@ -0,0 +1,61 @@
+        return statPerson?.details.count ?? 0
+    }
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if let cell = tableView.dequeueReusableCell(withIdentifier: "DetailCell",
```
Здесь хорошие заготовки для расширений, с помощью которых удобнее можно инициализировать и регистрировать ячейки, вьюхи и многое другое, пригодится: https://dev.to/tattn/my-favorite--swift-extensions-8g7

EugChesn replied:
Спасибище))

a-mkrv commented on TestView/Controllers/ExtensionsTable.swift:
```diff
@@ -0,0 +1,80 @@
+        return filteredData.count
+    }
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return 50
```
Вынеси все константы отдельно.

EugChesn replied:
fix

a-mkrv commented on TestView/Controllers/ExtensionsTable.swift:
```diff
@@ -0,0 +1,80 @@
+        startSpinner()
+        guard let info = timerSearchDelay?.userInfo as? String else { return }
+        if !info.isEmpty {
+            startSpinner()
```
Почему `startSpinner` два раза вызывается друг за другом?

EugChesn replied:
fix

a-mkrv commented on TestView/Controllers/ExtensionsTable.swift:
```diff
@@ -0,0 +1,80 @@
+        guard let info = timerSearchDelay?.userInfo as? String else { return }
+        if !info.isEmpty {
+            startSpinner()
+            model.searchPerson(search: info, completion: { [weak self] (data) in
```
`(data)` - скобки можно опустить, даже рекомендуется (как и во всех аналогичных местах)

EugChesn replied:
Так просто понятнее мне когда пишу closure)
Исправлю

EugChesn replied:
fix

a-mkrv commented on TestView/Model/ModelDataPerson.swift:
```diff
@@ -0,0 +1,155 @@
+import Reachability
+
+protocol SearchPerson: class {
+    func searchPerson(search: String,
```
Вместо двух замыканий, посмотри в сторону конструкции Result<T, Error>, где в качестве T - дженерик с типом, а Error - либо стандартная ошибка, либо можешь создавать свои кастомные ошибки. 
Будет что-то около того:
`searchPerson(search: String, completion: @escaping (Result<[String],NetworkError>) -> Void)
`

https://medium.com/@azamsharp/understanding-result-type-in-swift-5-387d5ef9f45e
https://www.hackingwithswift.com/articles/161/how-to-use-result-in-swift

EugChesn replied:
Fix)

a-mkrv commented on TestView/Model/ModelDataPerson.swift:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+    private func addObserverStateConnection() {
+        NotificationCenter.default.addObserver(self, selector: #selector(reachabilityChanged(note:)),
```
Каждый добавленный Обсервер необходимо "отписывать" в` deinit { } `перед тем, как будет закрыт экран. 

`NotificationCenter.default.removeObserver(self)`

EugChesn replied:
fix

a-mkrv commented on TestView/Model/ModelDataPerson.swift:
```diff
@@ -0,0 +1,155 @@
+    private func getPerson(recent: Set<String>) -> [String: ResultsStat] {
+        let keys = dataRequest.keys
+        var result: [String: ResultsStat] = [:]
+        for item in recent {
```
Как вариант можно писать так, либо через filters
```
for item in recent where keys.contains(item) {
    result[item] = dataRequest[item]
}
```

EugChesn replied:
fix

a-mkrv commented on TestView/Model/ModelDataPerson.swift:
```diff
@@ -0,0 +1,155 @@
+    }
+    private func addRequestIntoDictionary(request: SearchJson) {
+        if let result = request.results {
+            for person in result {
```
`result.forEach { dataRequest.updateValue($0, forKey: $0.name.lowercased()) }
`

https://habr.com/ru/post/440722/

EugChesn replied:
fix

a-mkrv commented on TestView/Model/ModelDataPerson.swift:
```diff
@@ -0,0 +1,155 @@
+
+    private func checkCashe(searchText: String) -> [String] {
+        var resultPersons: [String] = []
+        for key in dataRequest.keys {
```
Аналогично с` for ... where`

EugChesn replied:
fix

a-mkrv commented on TestView/Service/NetworkApi.swift:
```diff
@@ -0,0 +1,60 @@
+
+    private func createUrlSearch(namePeople: String) -> URL? {
+        let searchUrlStr = searchNameApi + namePeople
+        if let url = URL(string: searchUrlStr) {
```
У тебя же метод возвращает Optional, не обязательно проверять на `if let`
Можно сразу вернуть `URL(string: searchUrlStr)`

EugChesn replied:
fix

a-mkrv commented on TestView/UI/CustomCell/CustomDetailTableViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+class CustomDetailTableViewCell: UITableViewCell {
+    @IBOutlet var titleLabel: UILabel!
+    @IBOutlet var typeLabel: UILabel!
+    override func awakeFromNib() {
```
`awakeFromNib` можно удалить, он не используется никак, внутри нет никакой инициализации

EugChesn replied:
fix

a-mkrv commented on TestView/UI/CustomCell/CustomTableViewCell.swift:
```diff
@@ -0,0 +1,33 @@
+        super.setHighlighted(highlighted, animated: animated)
+        if highlighted {
+            contentView.backgroundColor = UIColor .systemGray
+            for item in self.subviews {
```
forEach

EugChesn replied:
fix

a-mkrv commented on TestView/UI/DataView/DataView.swift:
```diff
@@ -0,0 +1,32 @@
+    private func commonInit() {
+        let bundle = Bundle.init(for: DataView.self)
+        if let viewsToAdd = bundle.loadNibNamed("DataView", owner: self, options: nil),
+        let contentView = viewsToAdd.first as? UIView {
```
Посмотри ссылку (оставлял в этом PR в комментариях) про Extensions, которые инициализируют вьюхи удобно, избавит тебя от этого повторяющегося кода.

EugChesn committed:
Fix review)
1. Add extensions for table view register cells
2. Add completion: Result<Data,Error> instead success and fault
